File: Problems/No.936_Stamping_The_Sequence/stamping_the_sequence.py
class Solution(object):
    def movesToStamp(self, stamp, target):
        """
        :type stamp: str
        :type target: str
        :rtype: List[int]
        """

        # check if the target idx from i can be replaced by the stamp
        def Replacable(start_idx):
            for i in range(stamp_len):
                if ((target[start_idx + i] != stamp[i]) and (target[start_idx + i] != '?')):
                    return False
            return True


        # parameters
        stamp_idx = []
        stamp_len = len(stamp)
        target_len = len(target)
        replace_count = 0
        replaced = [False] * target_len
    
        # start computing
        while (replace_count < target_len):
            changed = False
            for i in range(target_len-stamp_len+1):
                if not replaced[i] and Replacable(i):
                    # replace the corresponding stamp in target to ? & count the replaced character num
                    changed = True
                    replaced[i] = True
                    for j in range(stamp_len):
                        if (target[i+j] != '?'):
                            replace_count += 1
                    target = target[:i] + ("?" * stamp_len) + target[i+stamp_len:]
                    stamp_idx.append(i)
                    if (replace_count == target_len):
                        break
            if not changed:
                return []
        
        return reversed(stamp_idx)

File: Problems/No.936_Stamping_The_Sequence/stamping_the_sequence.cpp
class Solution {
public:

    // check if the substring is all '?'
    bool AllQ(std::string &checking_str) {
       
        for (auto &c: checking_str) if (c != '?') return false;
        return true;

    }

    // check if the target idx from i can be replaced by the stamp
    bool Replacable(std::string &temp, std::string &stamp) {
       
        for (int i=0; i<stamp.length(); i++)
            if ((temp[i] != stamp[i]) && (temp[i] != '?'))
                return false;
        return true;

    }

    vector<int> movesToStamp(string stamp, string target) {
        
        // parameters
        std::vector<int> stamp_idx;
        int stamp_len = stamp.length();
        int target_len = target.length();
        bool changed = true;

        // start computing
        while (changed) {

            changed = false;
            for (int i=0; i<=target_len-stamp_len; i++) {

                std::string temp_str = target.substr(i, stamp_len);
                if (AllQ(temp_str)) continue;
                if (Replacable(temp_str, stamp)) {
             
                    // replace the corresponding stamp in target to ? & count the replaced character num
                    changed = true;
                    for (int j=0; j<stamp_len; j++) 
                        if (target[i+j] != '?') target[i+j] = '?';
                    stamp_idx.push_back(i);
                    
                }

            }

        }

        if (AllQ(target)) {

            std::reverse(stamp_idx.begin(), stamp_idx.end());
            return stamp_idx;

        }
        else return {};

    }
};
